test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java

```diff
@@ -0,0 +1,1226 @@
+    private boolean shouldVerifyIR; // Should we perform IR matching?
+    private static String lastTestVMOutput;
+
+    private final Class<?> testClass;
+    private List<Class<?>> helperClasses = null;
+    private List<Scenario> scenarios = null;
+    private final List<String> flags = new ArrayList<>();
+    private int defaultWarmup = -1;
+    private TestFrameworkSocket socket;
+    private Scenario scenario;
+
+    /*
+     * Public interface methods
+     */
+
+    /**
+     * Creates an instance acting as a builder to test the class from which this constructor was invoked from.
+     * Use this constructor if you want to use multiple run options (flags, helper classes, scenarios).
+     * Use the associated add methods ({@link #addFlags(String...)}, {@link #addScenarios(Scenario...)},
+     * {@link #addHelperClasses(Class...)}) to set up everything and then start the testing by invoking {@link #start()}.
+     */
+    public TestFramework() {
+        StackWalker walker = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE);
+        this.testClass = walker.getCallerClass();
+        if (VERBOSE) {
+            System.out.println("Test class: " + testClass);
+        }
+    }
+
+    /**
+     * Creates an instance acting as a builder to test {@code testClass}.
+     * Use this constructor if you want to use multiple run options (flags, helper classes, scenarios).
+     * Use the associated add methods ({@link #addFlags(String...)}, @link #addScenarios(Scenario...)},
+     * {@link #addHelperClasses(Class...)}) to set up everything and then start the testing by invoking {@link #start()}.
+     *
+     * @param testClass the class to be tested by the framework.
+     * @see #TestFramework()
+     */
+    public TestFramework(Class<?> testClass) {
+        TestRun.check(testClass != null, "Test class cannot be null");
+        this.testClass = testClass;
+        if (VERBOSE) {
+            System.out.println("Test class: " + testClass);
+        }
+    }
+
+    /**
+     * Default flags that are added used for the test VM.
+     */
+    private static String[] getDefaultFlags() {
+        return new String[] {"-XX:-BackgroundCompilation", "-XX:CompileCommand=quiet"};
+    }
+
+    /**
+     * Additional verification flags that are used if -DVerifyVM=true is with a debug build.
+     */
+    private static String[] getVerifyFlags() {
+        return new String[] {
+                "-XX:+UnlockDiagnosticVMOptions", "-XX:+VerifyOops", "-XX:+VerifyStack", "-XX:+VerifyLastFrame",
+                "-XX:+VerifyBeforeGC", "-XX:+VerifyAfterGC", "-XX:+VerifyDuringGC", "-XX:+VerifyAdapterSharing"
+        };
+    }
+
+    /**
+     * Tests the class from which this method was invoked from.
+     */
+    public static void run() {
+        StackWalker walker = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE);
+        run(walker.getCallerClass());
+    }
+
+    /**
```

Comment:
I'm not sure how much of the benefits all these different `run.*` bring. I fully understand the desire to simplify the most common use-case (i.e. no-arg `run`), but for the rest I'd assume the users will be capable of writing, for example, `new TestFramework(testClass).addScenarios(scenarios).start();` instead of `TestFramework .runWithScenarios(testClass, scenarios);`

Reply:
> I suggest to keep run() and maybe also runWithFlags() as (probably?) most common use-case.

I guess you are right, `run()` and `runWithFlags(String[])` will be most-commonly used ones.